src/vsr/replica.zig

```diff
@@ -4943,47 +4945,39 @@ pub fn ReplicaType(
         ///   ─/✓  op on disk at checkpoint
         ///   ·/%  op in memory at checkpoint
         ///     ✓  op_checkpoint
-        ///     %  op_checkpoint_trigger
+        ///     %  op_checkpoint's trigger
         ///
         fn op_checkpoint_next(self: *const Self) u64 {
+            assert(vsr.Op.checkpoint_valid(self.op_checkpoint()));
             assert(self.op_checkpoint() <= self.commit_min);
             assert(self.op_checkpoint() <= self.op or
                 self.status == .recovering or self.status == .recovering_head);
-            assert(self.op_checkpoint() == 0 or
-                (self.op_checkpoint() + 1) % constants.lsm_batch_multiple == 0);
-
-            const op = if (self.op_checkpoint() == 0)
-                // First wrap: op_checkpoint_next = 8-2-1 = 5
-                constants.journal_slot_count - constants.lsm_batch_multiple - 1
-            else
-                // Second wrap: op_checkpoint_next = 5+8-2 = 11
-                // Third wrap: op_checkpoint_next = 11+8-2 = 17
-                self.op_checkpoint() + constants.journal_slot_count - constants.lsm_batch_multiple;
-            assert((op + 1) % constants.lsm_batch_multiple == 0);
-            // The checkpoint always advances.
-            assert(op > self.op_checkpoint());
 
-            return op;
+            const checkpoint_next = vsr.Op.checkpoint_after(self.op_checkpoint());
+            assert(vsr.Op.checkpoint_valid(checkpoint_next));
+            assert(checkpoint_next > self.op_checkpoint()); // The checkpoint always advances.
+
+            return checkpoint_next;
         }
 
         /// Returns the next op that will trigger a checkpoint.
         ///
-        /// Receiving and storing an op higher than `op_checkpoint_trigger()` is forbidden; doing so
-        /// would overwrite a message (or the slot of a message) that has not yet been committed and
-        /// checkpointed.
+        /// Receiving and storing an op higher than `op_checkpoint_next_trigger()` is forbidden;
+        /// doing so would overwrite a message (or the slot of a message) that has not yet been
+        /// committed and checkpointed.
         ///
         /// See `op_checkpoint_next` for more detail.
-        fn op_checkpoint_trigger(self: *const Self) u64 {
-            return self.op_checkpoint_next() + constants.lsm_batch_multiple;
+        fn op_checkpoint_next_trigger(self: *const Self) u64 {
+            return vsr.Op.trigger_from_checkpoint(self.op_checkpoint_next()).?;
```

Comment:
maybe `trigger_for_checkpoint`? `for` is more specific about the relationship between the two.